File: day_6/firstChar.py
"""
  Non repeating character - 
  You are given a string consist of only lower case and upper case English 
  Alphabets and integers 0 to 9. Write a function that will take this string 
  as input and return the index of the first character that is non-repeating.
"""

# USING BRUTE FORCE

def find_first_non_repeating_char(str):
  for i in range(len(str)):
    flag = False
    for j in range(len(str)):
      if (str[i] == str[j] and i != j):
        flag = True
    
    if flag == False:
      return i
  
  return None


# UPTIMA METHOD

def firstNonRepeating(str):
  hashtable = {}

  for i in str:
    if i not in hashtable:
      hashtable[i] = 1
    else:
      hashtable[i] += 1

  for i in range(len(str)):
    if hashtable[str[i]] == 1:
      return i
  return None
  # return hashtable
  

a = 'a123412a'
print(find_first_non_repeating_char(a))
print(firstNonRepeating(a))

File: day_6/firstChar.js
/**
 * Non repeating character - 
 * You are given a string consist of only lower case and upper case English Alphabets 
 * and integers 0 to 9. Write a function that will take this string as input and return 
 * the index of the first character that is non-repeating. 
 */


// USING BRUTE FORCE

function nonRepeatingChar (str) {
  let flag;
  for (let i = 0; i < str.length; i++) {
    flag = false;
    for (let j = 0; j < str.length; j++) {
      if (str[i] === str[j] && i !== j ) {
        flag = true;
      } 
    }
    if ( flag === false) {
      return i;
    }
  }
  return null;
}


// USING HASH TABLE

function firstNonRepeating(str) {
  const hashTable = {};
  for (let i = 0; i< str.length; i++) {
    if(!hashTable[str[i]]) {
      hashTable[str[i]] = 1;
    } else {
      hashTable[str[i]] += 1;
    }
  }
  for (let i = 0; i< str.length; i++) {
    if (hashTable[str[i]] === 1) {
      return i;
    }
  }
  return null;
  // return hashTable;
}

function firstNonRepeatingChar (str) {
  const hashTable = {};
  for (let i = 0; i < str.length; i++) {
    if (str[i] in hashTable) {
      hashTable[str[i]] += 1;
    } else {
      hashTable[str[i]] = 1;
    }
  }
  for (let i = 0; i < str.length; i++) {
    if (hashTable[str[i]] === 1) return i;
  }
  return null;
}



const a = 'a123412a';
// const a = 'abcdbad';
console.log(nonRepeatingChar(a));
console.log(firstNonRepeating(a));
console.log(firstNonRepeatingChar(a));
// console.log(a.count('b'));